lib/pg_ha_migrations/unsafe_statements.rb

```diff
@@ -29,30 +29,45 @@ def self.delegate_unsafe_method_to_migration_base_class(method_name)
     ruby2_keywords "unsafe_#{method_name}"
   end
 
+  def self.delegate_unsafe_method_to_migration_base_class_with_lock(method_name, mode: :access_exclusive)
+    define_method("unsafe_#{method_name}") do |*args, &block|
+      if PgHaMigrations.config.check_for_dependent_objects
+        disallow_migration_method_if_dependent_objects!(method_name, arguments: args)
+      end
+
+      safely_acquire_lock_for_table(args.first, mode: mode) do
+        execute_ancestor_statement(method_name, *args, &block)
+      end
+    end
+    ruby2_keywords "unsafe_#{method_name}"
+  end
+
   def self.delegate_raw_method_to_migration_base_class(method_name)
     define_method("raw_#{method_name}") do |*args, &block|
       execute_ancestor_statement(method_name, *args, &block)
     end
     ruby2_keywords "raw_#{method_name}"
   end
 
-  # Direct dispatch to underlying Rails method as unsafe_<method_name> with dependent object check
-  delegate_unsafe_method_to_migration_base_class :add_check_constraint
-  delegate_unsafe_method_to_migration_base_class :add_column
-  delegate_unsafe_method_to_migration_base_class :add_foreign_key
-  delegate_unsafe_method_to_migration_base_class :change_column
-  delegate_unsafe_method_to_migration_base_class :change_column_default
-  delegate_unsafe_method_to_migration_base_class :change_table
+  # Direct dispatch to underlying Rails method as unsafe_<method_name> with dependent object check / safe lock acquisition
+  delegate_unsafe_method_to_migration_base_class_with_lock :add_check_constraint
+  delegate_unsafe_method_to_migration_base_class_with_lock :add_column
+  delegate_unsafe_method_to_migration_base_class_with_lock :change_column
```

Comment:
It seems like we should treat this like we do `change_table`?

Reply:
> Like if you wanted to change a column type would you be forced to use `raw_change_column`?

Hmm, that's a good point. We probably should have a way to do those things discretely (and that would allow a safe version for types, also, since some are binary compatible).

But the non-existence of those now suggests we can punt that to an issue to track it.

Reply:
Yeah I was thinking about that a bit too. I can open an issue and we can discuss further, but I'm assuming the list of binary compatible types needs to be whittled down to type changes that Rails can transparently handle. For example, json -> text appears to be binary compatible, but I imagine Rails might not like that

Reply:
Yeah, sounds good.

Reply:
https://github.com/braintree/pg_ha_migrations/issues/128